File: data-structures/array/array-class/java/Array.java
import java.lang.IndexOutOfBoundsException;

/**
 * The {@code Array} class represents a dynamic array data structure that can
 * store integers.
 * It provides methods for appending values, removing values at specified
 * indices,
 * retrieving values by index, and finding the index of a value.
 */
public class Array {
    private int size;
    private int capacity;
    private int[] values;

    /**
     * Constructs a new Array object with the specified capacity.
     *
     * @param capacity the initial capacity of the array
     */
    public Array(int capacity) {
        this.size = 0;
        this.capacity = capacity;
        this.values = new int[capacity];
    }

    /**
     * Returns the current size of the array.
     *
     * @return the size of the array
     */
    public int getSize() {
        return this.size;
    }

    /**
     * Appends the specified value to the end of the array.
     * If the array is full, it doubles its capacity before appending the value.
     *
     * @param value the value to be appended
     */
    public void append(int value) {
        if (this.capacity == this.size) {
            int[] tempValues = new int[this.capacity * 2];
            System.arraycopy(this.values, 0, tempValues, 0, this.size);
            this.values = tempValues;
            this.capacity *= 2;
        }
        this.values[this.size] = value;
        this.size += 1;
    }

    /**
     * Removes the value at the specified index from the array.
     * If the index is out of range, it prints an error message and does nothing.
     *
     * @param index the index of the value to be removed
     */
    public void removeAt(int index) {
        if (index >= this.size || index < 0) {
            throw new IndexOutOfBoundsException("Index out of range");
        }
        if (index == this.size - 1) {
            this.size -= 1;
            return;
        }

        System.arraycopy(this.values, index + 1, this.values, index, this.size - index - 1);
        this.size -= 1;
    }

    /**
     * Retrieves the value at the specified index from the array.
     * If the index is out of range, it returns -1.
     *
     * @param index the index of the value to be retrieved
     * @return the value at the specified index, or -1 if the index is out of range
     */
    public int get(int index) {
        if (index >= this.size || index < 0) {
            throw new IndexOutOfBoundsException("Index out of range");
        }

        return this.values[index];
    }

    /**
     * Returns the index of the first occurrence of the specified value in the
     * array.
     * If the value is not found, it returns -1.
     *
     * @param value the value to be searched
     * @return the index of the value, or -1 if the value is not found
     */
    public int indexOf(int value) {
        for (int i = 0; i < this.size; i++) {
            if (values[i] == value) {
                return i;
            }
        }
        return -1;
    }

}

File: data-structures/array/array-class/python/array.py


import unittest


class Array:
    """
    An array implementation with a fixed capacity.

    Args:
        capacity (int): The maximum number of elements the array can hold.

    Attributes:
        _capacity (int): The maximum capacity of the array.
        _size (int): The current number of elements in the array.
        _values (list): The list storing the elements of the array.

    Methods:
        get_size() -> int:
            Returns the current number of elements in the array.

        append(value: int) -> None:
            Appends the given value to the end of the array.

        remove_at(index: int) -> None:
            Removes the element at the specified index from the array.

        get(index: int) -> int:
            Returns the element at the specified index.

        indexOf(value: int) -> int:
            Returns the index of the first occurrence of the specified value in the array. Returns -1 if not found.
    """

    def __init__(self, capacity: int) -> None:
        """
        Initializes a new instance of the Array class with the specified capacity.

        Args:
            capacity (int): The maximum number of elements the array can hold.

        Attributes:
            _capacity (int): The maximum capacity of the array.
            _size (int): The current number of elements in the array.
            _values (list): The list storing the elements of the array.
        """
        self._capacity = capacity
        self._size = 0
        self._values = [None for i in range(self._capacity)]

    def get_size(self) -> int:
        """
        Returns the current number of elements in the array.

        Returns:
            int: The number of elements in the array.
        """
        return self._size

    def append(self, value: int) -> None:
        """
        Appends the given value to the end of the array.

        Args:
            value (int): The value to be appended to the array.

        Raises:
            ValueError: If the value is not an integer.
        """
        if not isinstance(value, int):
            raise ValueError(f"{value} must be an int")
        if self._size == self._capacity:
            self._values.extend([None for i in range(self._capacity)])
        self._values[self._size] = value
        self._size += 1

    def remove_at(self, index: int) -> None:
        """
        Removes the element at the specified index from the array.

        Args:
            index (int): The index of the element to be removed.

        Raises:
            ValueError: If the index is out of range.
        """
        if index < 0 or index >= self._size:
            raise ValueError("Index not found")
        if (index == self._size - 1):
            self._size -= 1
            return
        for i in range(index, self._size):
            self._values[i] = self._values[i+1]
        self._size -= 1

    def get(self, index: int) -> int:
        """
        Returns the element at the specified index.

        Args:
            index (int): The index of the element to be retrieved.

        Returns:
            int: The value at the specified index.

        Raises:
            ValueError: If the index is out of range.
        """
        if index < 0 or index >= self._size:
            raise ValueError("Index not found")
        return self._values[index]

    def indexOf(self, value: int) -> None:
        """
        Returns the index of the first occurrence of the specified value in the array.

        Args:
            value (int): The value to search for in the array.

        Returns:
            int: The index of the first occurrence of the value, or -1 if not found.

        Raises:
            ValueError: If the value is not an integer.
        """
        if not isinstance(value, int):
            raise ValueError(f"{value} must be an int")
        for i in range(self._size):
            if value == self._values[i]:
                return i
        return -1
